### DISCUSSION TITLE
Understanding the way 3D terrain is incorporated though the wind potential $\Phi$

### DISCUSSION THREAD

**Discussion by alexr314** [firelab/windninja#534] [General]
Hello WindNinja team,

I'm Dr. Alex Roman, I'm a physicist and ML consultant, and I'm working with Improving Aviation on using wind models to predict ember transport. I’m trying to understand in detail how WindNinja works.

Specifically, I'm trying to understand how WindNinja uses a scalar "wind potential" field to model the 3D effects of terrain.
As I understand it, WindNinja solves a 2D mass-consistent fluid model to find the 3D velocity vector (u_x, u_y, u_z) everywhere on a 2D surface a fixed distance above the surface of the terrain. What I have just described does not yet have any relation to the 3D topography, so WindNinja initially solves for the wind potential Phi using this equation: (borrowed from the source code:)

```
    d        dPhi      d        dPhi      d        dPhi
   ---- ( Rx ---- ) + ---- ( Ry ---- ) + ---- ( Rz ---- ) + H = 0.0
    dx        dx       dy        dy       dz        dz
```

With boundary conditions:

- Flow through => Phi = 0       (Dirichlet)
- Ground => normal flux = 0   (Neumann)

(The ground condition makes sense, I don’t understand what the Dirichlet condition really means. I’m also uncertain about what boundary condition is used for the top, I assume it’s also Phi=0)

This requires us to have computed H. I’m unclear on what H is, I see that it is calculated in the source code as the divergence of the initial field

```
         du0     dv0     dw0
    H = ----- + ----- + -----
         dx      dy      dz
```

This relies upon the initial wind vector field, which is sometimes just a uniform field having which would make H uniformly 0. In this case the PDE for Phi would just be the Laplace equation, and with these boundary conditions I can’t see why the solution would be nontrivial. If we computed Phi and it was trivial, then I don’t see how the simulation uses any information about the topography.

In summary:
- The user either inputs a single wind velocity (or a set of velocities which are averaged) and used for the entire domain ("domain averaging") or the user provided a list of weather station measurements ("point initialization") and the input is somehow constructed from them?
- the initial wind vector field allows you to compute H, 
- H along with the topography allow you to compute Phi, 
- the gradient of Phi updates the initial wind vectors (u0,v0,w0) to compute (u,v,w)
- Then what? Is this the final output?

Please let me know in what ways this is not an accurate description of the method or what complexities I am missing.

My questions are:
- How is the initial wind velocity field determined if the user chose "point initialization"?
- If the user chose "domain averaging" why is H not just identically 0?
- What is the boundary condition for Phi at the top of the mesh (high altitude)?
- Does the topographical data impact the result in any other way than through Phi?
- At which stage in this process exactly does the fluid solving happen?

Pardon me if the answers to any of these questions are already provided in some obvious way. I look forward to hearing from you.
Thanks in advance for your answers! 
 -- Alex

**Comment by nwagenbrenner** [comment]
@alexr314 Thanks for the questions. 

WindNinja enforces conservation of mass on a 3-D grid. The model is described in detail in this paper:

https://weather.firelab.org/windninja/publications/comparison.pdf

The terrain is accounted for in the finite element method used to solve the governing equations on the 3-D grid. A "flow-through" or "open" boundary condition is used for all non-ground boundaries (the top and lateral boundaries of the domain) by setting Phi = 0. The H is like a source term that is a function of the divergence of the initial wind field. The initial wind field (denoted u0, v0, w0 in the source code) is a 3-D wind field that can come from a coarser weather model, point observations, or a user-specified domain-average wind field. In each of these cases, we use a log profile to fill the WindNinja 3-D grid vertically and interpolation to fill the grid horizontally for coarse weather model or point initializations. So the initial field is never fully uniform (there is always a vertical gradient). The three initialization options are referred to as "weather model initialization," "point initialization", and "domain average initialization" in the code.

I hope this helps. Please let us know if you have additional questions or want more detail.

**Reply by alexr314** [reply]
Thank you for the thoughtful response. This helps a great deal, I worked through the paper and now understand much better. I hadn't appreciated the way that the logarithmic profile is evaluated with respect to the height above ground level incorporating topography naturally. The whole setup is very cool!

The only thing I'm confused about at the moment is why want Phi=0 on the domain boundary?
Could we equally well require zero normal derivative? (Though, I realize that in this case nothing constrains the actual value of Phi). Or could we require all components of the gradient of Phi to be 0 at the boundaries? (Perhaps this is too restrictive? It seems like there might not be guaranteed a solution. But this does cause it to fit the interpolated ambient field perfectly).

Thank you!

**Reply by alexr314** [reply]
I just realized that my proposed boundary conditions would enforce a total flux of 0 across the boundary, and the divergence theorem would require the net divergence in the domain to be 0 which isn't usually true. Is this the way you think of it? The Dirichlet condition is the most restrictive condition which still allows for arbitrary flux?

**Reply by nwagenbrenner** [reply]
@alexr314 Sorry for the slow reply. Yeah, that's right - we need to allow for a flux across the flow through boundaries (the flux across the flow through boundaries is not necessarily 0). Since PHI is like a potential flow, setting PHI = 0 at the flow through boundaries is like saying the potential flow is 0 at the boundary (in the normal direction). I guess this is similar to a zero gradient boundary condition.
